cloud-common/src/main/java/com/sequenceiq/cloudbreak/util/FixedSizePreloadCache.java

```diff
@@ -0,0 +1,42 @@
+package com.sequenceiq.cloudbreak.util;
+
+import java.util.Stack;
+import java.util.function.Supplier;
+import java.util.stream.IntStream;
+
+public class FixedSizePreloadCache<E> {
+
+    private final int maxSize;
+
+    private final Stack<E> stack;
+
+    private final Supplier<E> generator;
+
+    public FixedSizePreloadCache(int maxSize, Supplier<E> generator) {
+        stack = new Stack<>();
+        this.maxSize = maxSize;
+        this.generator = generator;
+
+        IntStream.range(0, maxSize).parallel().forEach(i -> push(generator.get()));
+    }
+
+    public E pop() {
+        E element;
+        synchronized (stack) {
```

Comment:
Instead of synchronising an inherently concurrent collection should be used. e.g. https://docs.oracle.com/javase/10/docs/api/java/util/concurrent/ConcurrentLinkedQueue.html

Reply:
I chose `synchronization` because i don't want to loose control over the number of elements. This class is useful for heavy objects, where synchronization cost is nothing. Harder to guarantee the number of elements with any other non-blocking algorithm.